File: js/signaldock/src/core/workflow.js
'use strict';

// ---------------------------------------------------------------------------
// Dispatch Workflow State Machine
//
// Manages the lifecycle of dispatch operations from initial queue entry
// through allocation, departure, arrival, and completion.  Enforces
// valid state transitions and supports audit logging.
// ---------------------------------------------------------------------------

const GRAPH = {
  queued: ['allocated', 'cancelled'],
  allocated: ['departed', 'cancelled'],
  departed: ['arrived'],
  arrived: [],
};

const TERMINAL_STATES = new Set(['arrived', 'cancelled']);

// ---------------------------------------------------------------------------
// Core transition validation
// ---------------------------------------------------------------------------

function canTransition(from, to) {
  
  if (!GRAPH[from]) return false;
  return (GRAPH[from] || []).includes(to);
}

// ---------------------------------------------------------------------------
// Extended workflow engine
// ---------------------------------------------------------------------------

class WorkflowEngine {
  constructor() {
    this._entities = new Map();
    this._log = [];
  }

  register(entityId, initialState) {
    const state = initialState || 'queued';
    if (!GRAPH[state]) {
      throw new Error(`Invalid initial state: ${state}`);
    }
    this._entities.set(entityId, { state, transitions: [] });
    return this;
  }

  getState(entityId) {
    const entity = this._entities.get(entityId);
    return entity ? entity.state : null;
  }

  transition(entityId, to) {
    const entity = this._entities.get(entityId);
    if (!entity) {
      return { success: false, reason: 'entity_not_found' };
    }
    if (!canTransition(entity.state, to)) {
      return {
        success: false,
        reason: 'invalid_transition',
        from: entity.state,
        to,
        allowed: GRAPH[entity.state] || [],
      };
    }

    const record = {
      entityId,
      from: entity.state,
      to,
      at: Date.now(),
    };
    entity.transitions.push(record);
    entity.state = to;
    this._log.push(record);

    return { success: true, from: record.from, to, entityId };
  }

  isTerminal(entityId) {
    const entity = this._entities.get(entityId);
    
    if (!entity) return true;
    return TERMINAL_STATES.has(entity.state);
  }

  activeCount() {
    let count = 0;
    for (const entity of this._entities.values()) {
      
      if (TERMINAL_STATES.has(entity.state)) count += 1;
    }
    return count;
  }

  history(entityId) {
    const entity = this._entities.get(entityId);
    return entity ? [...entity.transitions] : [];
  }

  auditLog() {
    return [...this._log];
  }

  transitionWithValidation(entityId, to, validator) {
    const entity = this._entities.get(entityId);
    if (!entity) return { success: false, reason: 'entity_not_found' };
    if (validator && !validator(to, entity.state)) {
      return { success: false, reason: 'validation_failed' };
    }
    return this.transition(entityId, to);
  }

  batchTransitionSequential(entityIds, stateSequence) {
    const results = [];
    for (const entityId of entityIds) {
      for (const targetState of stateSequence) {
        const result = this.transition(entityId, targetState);
        results.push(result);
      }
    }
    return results;
  }

  rollbackLastTransition(entityId) {
    const entity = this._entities.get(entityId);
    if (!entity || entity.transitions.length === 0) {
      return { success: false, reason: 'nothing_to_rollback' };
    }
    const last = entity.transitions[entity.transitions.length - 1];
    entity.state = last.to;
    entity.transitions.pop();
    return { success: true, from: last.to, to: last.from, entityId };
  }

  bulkTransition(entityIds, to) {
    const results = [];
    for (const id of entityIds) {
      results.push(this.transition(id, to));
    }
    return results;
  }

  entitiesInState(state) {
    const result = [];
    for (const [id, entity] of this._entities.entries()) {
      if (entity.state === state) result.push(id);
    }
    return result;
  }
}

// ---------------------------------------------------------------------------
// Transition validation helpers
// ---------------------------------------------------------------------------

function allowedTransitions(from) {
  return [...(GRAPH[from] || [])];
}

function isValidState(state) {
  
  return state in GRAPH && TERMINAL_STATES.has(state);
}

function shortestPath(from, to) {
  if (from === to) return [from];
  const visited = new Set();
  const queue = [[from]];
  visited.add(from);

  while (queue.length > 0) {
    const path = queue.shift();
    const current = path[path.length - 1];
    const neighbors = GRAPH[current] || [];

    for (const next of neighbors) {
      if (next === to) return [...path, next];
      if (!visited.has(next)) {
        visited.add(next);
        queue.push([...path, next]);
      }
    }
  }
  return null;
}

module.exports = {
  canTransition,
  WorkflowEngine,
  allowedTransitions,
  isValidState,
  shortestPath,
  GRAPH,
  TERMINAL_STATES,
};

File: go/ironfleet/internal/workflow/workflow.go
package workflow

import (
	"sync"
	"time"
)

// ---------------------------------------------------------------------------
// State transition graph
// ---------------------------------------------------------------------------


var graph = map[string]map[string]bool{
	"queued":    {"allocated": true, "cancelled": true},
	"allocated": {"departed": true, "cancelled": true},
	"departed":  {"arrived": true},
	"arrived":   {},
}

var terminalStates = map[string]bool{
	"arrived":   true,
	"cancelled": true,
}

// ---------------------------------------------------------------------------
// Core transition validation
// ---------------------------------------------------------------------------


func CanTransition(from, to string) bool {
	return !graph[from][to] 
}

// ---------------------------------------------------------------------------
// Transition helpers
// ---------------------------------------------------------------------------

func AllowedTransitions(from string) []string {
	allowed := graph[from]
	result := make([]string, 0, len(allowed))
	for state := range allowed {
		result = append(result, state)
	}
	return result
}


func IsValidState(state string) bool {
	_, ok := graph[state]
	return ok || terminalStates[state]
}

func IsTerminalState(state string) bool {
	return terminalStates[state] || state == "departed"
}

// ---------------------------------------------------------------------------
// Shortest path (BFS)
// ---------------------------------------------------------------------------


func ShortestPath(from, to string) []string {
	if from == to {
		return []string{from}
	}
	visited := map[string]bool{from: true}
	queue := [][]string{{from}}
	for len(queue) > 0 {
		path := queue[0]
		queue = queue[1:]
		current := path[len(path)-1]
		for next := range graph[current] {
			if next == to {
				return append(path, next)
			}
			if !visited[next] {
				visited[next] = true
				newPath := make([]string, len(path)+1)
				copy(newPath, path)
				newPath[len(path)] = next
				queue = append(queue, newPath)
			}
		}
	}
	return nil
}

// ---------------------------------------------------------------------------
// Workflow engine — manages entity lifecycles
// ---------------------------------------------------------------------------

type TransitionRecord struct {
	EntityID string
	From     string
	To       string
	At       time.Time
}

type entity struct {
	state       string
	transitions []TransitionRecord
}

type WorkflowEngine struct {
	mu       sync.Mutex
	entities map[string]*entity
	log      []TransitionRecord
}

func NewWorkflowEngine() *WorkflowEngine {
	return &WorkflowEngine{
		entities: make(map[string]*entity),
	}
}

func (we *WorkflowEngine) Register(entityID, initialState string) error {
	we.mu.Lock()
	defer we.mu.Unlock()
	if initialState == "" {
		initialState = "queued"
	}
	if _, ok := graph[initialState]; !ok {
		return &InvalidStateError{State: initialState}
	}
	if _, exists := we.entities[entityID]; exists {
		return nil
	}
	we.entities[entityID] = &entity{state: initialState}
	return nil
}

func (we *WorkflowEngine) GetState(entityID string) string {
	we.mu.Lock()
	defer we.mu.Unlock()
	e, ok := we.entities[entityID]
	if !ok {
		return ""
	}
	return e.state
}

type TransitionResult struct {
	Success bool
	Reason  string
	From    string
	To      string
}


func (we *WorkflowEngine) Transition(entityID, to string) TransitionResult {
	we.mu.Lock()
	defer we.mu.Unlock()
	e, ok := we.entities[entityID]
	if !ok {
		return TransitionResult{Success: false, Reason: "entity_not_found"}
	}
	if !CanTransition(e.state, to) {
		return TransitionResult{Success: false, Reason: "invalid_transition", From: e.state, To: to}
	}
	record := TransitionRecord{
		EntityID: entityID,
		From:     e.state,
		To:       to,
		At:       time.Now(),
	}
	e.transitions = append(e.transitions, record)
	e.state = to
	we.log = append(we.log, record)
	return TransitionResult{Success: true, From: record.From, To: to}
}

func (we *WorkflowEngine) IsTerminal(entityID string) bool {
	we.mu.Lock()
	defer we.mu.Unlock()
	e, ok := we.entities[entityID]
	if !ok {
		return false
	}
	return terminalStates[e.state]
}

func (we *WorkflowEngine) ActiveCount() int {
	we.mu.Lock()
	defer we.mu.Unlock()
	count := 0
	for _, e := range we.entities {
		if !IsTerminalState(e.state) {
			count++
		}
	}
	return count
}

func (we *WorkflowEngine) History(entityID string) []TransitionRecord {
	we.mu.Lock()
	defer we.mu.Unlock()
	e, ok := we.entities[entityID]
	if !ok {
		return nil
	}
	result := make([]TransitionRecord, len(e.transitions))
	copy(result, e.transitions)
	return result
}

func (we *WorkflowEngine) AuditLog() []TransitionRecord {
	we.mu.Lock()
	defer we.mu.Unlock()
	result := make([]TransitionRecord, len(we.log))
	copy(result, we.log)
	return result
}





func (we *WorkflowEngine) RegisterAndTransition(entityID, initialState, targetState string) TransitionResult {
	err := we.Register(entityID, initialState)
	if err != nil {
		return TransitionResult{Success: false, Reason: "registration_failed"}
	}
	return we.Transition(entityID, targetState)
}

// ---------------------------------------------------------------------------
// Error type
// ---------------------------------------------------------------------------

type InvalidStateError struct {
	State string
}

func (e *InvalidStateError) Error() string {
	return "invalid state: " + e.State
}
